wandb/integration/kfp/kfp_patch.py

```diff
@@ -0,0 +1,309 @@
+        )
+
+except ImportError:
+    wandb.termerror("kfp not found!  Please `pip install kfp`")
+
+from .wandb_logging import wandb_log
+
+decorator_code = inspect.getsource(wandb_log)
+wandb_logging_extras = f"""
+import typing
+from typing import NamedTuple
+
+import collections
+from collections import namedtuple
+
+import kfp
+from kfp import components
+from kfp.components import InputPath, OutputPath
+
+import wandb
+
+{decorator_code}
+"""
+
+
+def full_path_exists(full_func):
+    def get_parent_child_pairs(full_func):
+        components = full_func.split(".")
+        parents, children = [], []
+        for i, _ in enumerate(components[:-1], 1):
+            parent = ".".join(components[:i])
+            child = components[i]
+            parents.append(parent)
+            children.append(child)
+        return zip(parents, children)
+
+    for parent, child in get_parent_child_pairs(full_func):
+        module = wandb.util.get_module(parent)
+        if not module or not hasattr(module, child) or getattr(module, child) is None:
+            return False
+    return True
+
+
+def patch(module_name, func):
+    module = wandb.util.get_module(module_name)
+    success = False
+
+    full_func = f"{module_name}.{func.__name__}"
+    if not full_path_exists(full_func):
+        wandb.termerror(
+            f"Failed to patch {module_name}.{func.__name__}!  Please check if this package/module is installed!"
+        )
+    else:
+        wandb.patched.setdefault(module.__name__, [])
+        # if already patched, do not patch again
+        if [module, func.__name__] not in wandb.patched[module.__name__]:
+            setattr(module, f"orig_{func.__name__}", getattr(module, func.__name__))
+            setattr(module, func.__name__, func)
+            wandb.patched[module.__name__].append([module, func.__name__])
+        success = True
+
+    return success
+
+
+def unpatch(module_name):
+    if module_name in wandb.patched:
+        for module, func in wandb.patched[module_name]:
+            setattr(module, func, getattr(module, f"orig_{func}"))
+        wandb.patched[module_name] = []
+
+
+def unpatch_kfp():
+    unpatch("kfp.components")
+    unpatch("kfp.components._python_op")
+    unpatch("wandb.integration.kfp")
+
+
+def patch_kfp():
+    to_patch = [
+        ("kfp.components", create_component_from_func,),
+        ("kfp.components._python_op", create_component_from_func,),
+        ("kfp.components._python_op", _get_function_source_definition,),
+        ("kfp.components._python_op", strip_type_hints),
+    ]
+
+    successes = []
+    for module_name, func in to_patch:
+        success = patch(module_name, func)
+        successes.append(success)
+    if not all(successes):
+        wandb.termerror(
+            "Failed to patch one or more kfp functions.  Patching @wandb_log decorator to no-op."
+        )
+        patch("wandb.integration.kfp", wandb_log)
```

Comment:
Hmm, if we were unable to import these functions, calling patch_kfp would blow up with undefined errors.  Not sure if we're already handling that case but it should be easy to set a boolean in the ImportError case and make methods like this print a warning.